utils/src/metrics.rs

```diff
@@ -0,0 +1,98 @@
+use prometheus_client::encoding::EncodeLabelSet;
+use prometheus_client::metrics::counter::Counter;
+use prometheus_client::metrics::family::Family;
+
+/// Metric label that indicates success or failure.
+#[derive(Clone, Debug, Hash, PartialEq, Eq, EncodeLabelSet)]
+pub struct Status {
+    pub ok: &'static str,
+}
+
+/// Success and failure labels.
+const SUCCESS: Status = Status { ok: "o" };
+const FAILURE: Status = Status { ok: "x" };
+
+/// A counter metric with a boolean status label indicating success or failure.
+pub type StatusCounter = Family<Status, Counter>;
+
+/// Trait providing convenience methods for `StatusCounter`.
+pub trait StatusExt {
+    fn inc_ok(&self);
+    fn inc_fail(&self);
+}
+
+impl StatusExt for StatusCounter {
+    /// Increment the metric as a success.
+    fn inc_ok(&self) {
+        self.get_or_create(&SUCCESS).inc();
+    }
+
+    /// Increment the metric as a failure.
+    fn inc_fail(&self) {
+        self.get_or_create(&FAILURE).inc();
+    }
+}
+
+/// Records metrics when dropped.
+///
+/// Records can be used to ensure that certain metrics are incremented regardless of the control
+/// flow. For example, if a function returns early, the metric will still be incremented.
+pub struct Record {
+    /// The metric to record.
+    metric: Family<Status, Counter>,
+
+    /// Whether the record will increment the metric as a success or failure.
+    ok: bool,
+}
+
+impl Record {
+    /// Create a new record.
+    ///
+    /// The record is marked as a failure by default. This is to ensure that the metric is
+    /// incremented as a failure if the record is dropped early (e.g. due to an error).
+    ///
+    /// To default to success, use [`Record::ok`], e.g.:
+    /// ```
+    /// let _record = Record::new(&metric).ok();
+    /// ```
+    pub fn new(metric: &Family<Status, Counter>) -> Self {
```

Comment:
I think it would be cleaner to implement a function on `StatusCounter` (what I think you should just call `Status`) that gives you a new "instance".